### PR TITLE
Expand operations to support moving families in and out of tanks

### PR DESCRIPTION
**Pull request by zspencer**
# Checklist:

- [x]  I have performed a self-review of my own code,
- [x] I have commented my code, particularly in hard-to-understand areas,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works,
- [x] New and existing unit tests pass locally with my changes ("bundle exec rake"),
- [x] Title include "WIP" if work is in progress.

See: #139 

### Type of change

* New feature (non-breaking change which adds functionality)

### How Has This Been Tested?

Tested by running the rspec tests, there does not appear to be a UI or other interface for performing Operations.

### CODE REVIEW

**Comment by zspencer** [comment]
@mdworken - Hey Michael!

If you happen to merge #147 before this, I will rebase it out of this branch; or we can merge this branch as is and close #147; since it's an incidental refactor.

**Comment by mdworken** [comment]
@zspencer I'm so sorry, saw this one second - could you rebase if you don't mind?

**Comment by zspencer** [comment]
@mdworken - Rebase complete!

File: app/models/operation.rb [hunk 1]
```diff
@@ -1,3 +1,25 @@
 class Operation < ApplicationRecord
   belongs_to :tank
+  belongs_to :family, required: false
+
+  validates :family, presence: true, if: :add_family?
```
Comment (zspencer) [inline app/models/operation.rb:5]: @jtu0 / @mdworken - Thoughts on this being `if: :needs_family?` to more clearly express intent?

File: app/models/operation.rb [hunk 2]
```diff
@@ -1,3 +1,25 @@
...
+
+  def perform
+    case action.to_sym
```
Comment (zspencer) [inline app/models/operation.rb:8]: We call `action&.to_sym` a few different places, we may want to extract the following method:

```
def action
  action&.to_sym
end
```

File: db/seeds.rb [hunk 3]
```diff
@@ -64,7 +64,7 @@
 female = Animal.create!(sex: 'female')
 family = Family.create!(male: male, female: female)
 tank = Tank.create!(facility: Facility.find_by(code: 'PSRF'), name: 'AB-17')
-Operation.create!(tank: tank, animals_added: 800, animals_added_family_id: family.id, operation_date: 7.days.ago, action: 'add')
+Operation.create!(tank: tank, animals_added: 800, family: family, animals_added_family_id: family.id, operation_date: 7.days.ago, action: 'add')
```
Comment (zspencer) [inline db/seeds.rb:67]: @mdworken  - Could you clarify what the purpose of the `animals_added_family_id` is supposed to signify? Is this leftover that can be removed?

File: spec/models/operation_spec.rb [hunk 4]
```diff
@@ -1,7 +1,67 @@
...
+      # syntax does not always play nicely.
+      #
+      # We wrap operation.perform call in a proc so that the lambda expression
+      # can be evaluated by the raise_error block.
+      #
+      # This is part of why some codebases prefer to avoid the "it { ... }" syntax
+      # in favor of "it 'words' do ... end" syntax.
+      subject(:proc) { -> { operation.perform } }
+      let(:action) { :eat_family }
+      it { is_expected.to raise_error(Operation::InvalidActionError) }
```
Comment (zspencer) [inline spec/models/operation_spec.rb:64]: Another way we could do this without the gross `subject(:proc) { -> { operation.perform } }` would be:

```
it "raises an error" do
  expect { operation.perform }.to raise_error(Operation::InvalidActionError)
end
```

Which is where I _think_ @jtu0 was trying to get us to during the pairing session but I couldn't see the shape of it until now.

File: db/seeds.rb (hunk 3 above)
Comment (mdworken) [reply db/seeds.rb:67]: I think it probably is, yes!

File: spec/models/operation_spec.rb (hunk 4 above)
Comment (mdworken) [reply spec/models/operation_spec.rb:64]: Eh, I'm fine either way on this honestly.

**Pull request merged**